src/riak_core_worker_pool.erl

```diff
@@ -162,68 +153,48 @@ shutdown({work, _Work, From}, State) ->
 shutdown(_Event, State) ->
     {next_state, shutdown, State}.
 
-handle_event({checkin, Pid},
-                shutdown,
-                #state{pool=Pool,
-                        pool_name=PoolName,
-                        monitors=Monitors0,
-                        checkouts=Checkouts0} = State) ->
-    Monitors = demonitor_worker(Pid, Monitors0),
+handle_event({checkin, Worker}, StateName, State) ->
+    Monitors = demonitor_worker(Worker, State#state.monitors),
     {ok, Checkouts} =
-        poolboy_checkin(Pool, Pid, PoolName, Checkouts0),
-    case Monitors of
-        [] -> %% work all done, time to exit!
+        poolboy_checkin(
+            State#state.pool,
+            Worker,
+            State#state.pool_name,
+            State#state.checkouts
+    ),
+    case {StateName, Monitors} of
+        {shutdown, []} ->
             {stop, shutdown, State};
-        _ ->
+        {shutdown, _} ->
             {next_state,
                 shutdown,
-                State#state{monitors=Monitors, checkouts=Checkouts}}
-    end;
-handle_event({checkin, Worker},
-                _,
-                #state{pool=Pool,
-                        queue=Q,
-                        pool_name=PoolName,
-                        checkouts=Checkouts0,
-                        monitors=Monitors0} = State) ->
-    case consume_from_queue(Q, State#state.pool_name) of
-        {{value, {work, Work, From}}, Rem} ->
-            %% there is outstanding work to do - instead of checking
-            %% the worker back in, just hand it more work to do
-            Monitors = monitor_worker(Worker, From, Work, Monitors0),
-            do_work(Worker, Work, From, State#state.callback_mod),
-            {next_state,
-                queueing, State#state{queue=Rem, monitors=Monitors}};
-        {empty, Empty} ->
-            Monitors = demonitor_worker(Worker, Monitors0),
-            {ok, Checkouts} =
-                poolboy_checkin(Pool, Worker, PoolName, Checkouts0),
-            {next_state,
-                ready,
-                State#state{queue=Empty,
-                            monitors=Monitors,
-                            checkouts=Checkouts}}
+                State#state{monitors=Monitors, checkouts=Checkouts}};
+        _ ->
```

Comment:
This combines the early top level split of checkin in `shutdown` state or any other state.

But there is a subtle difference. In the original implementation, one could send an event `{checkin, Pid}` where the `Pid` is not part of the checkouts and then that pid would get monitored straightaway, whereas in the new implementation, that monitoring has to wait until the checkouts are empty... at least that's what I read in the code.